presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -513,19 +619,39 @@ else if (taskState == TaskState.FINISHED) {
         }
     }
 
+    private boolean isRecoverable(List<ExecutionFailureInfo> failures)
+    {
+        for (ExecutionFailureInfo failure : failures) {
+            if (!RECOVERABLE_ERROR_CODES.contains(failure.getErrorCode())) {
+                return false;
+            }
+        }
+        return stageTaskRecoveryCallback.isPresent() &&
+                failedTasks.size() < allTasks.size() * maxFailedTaskPercentage;
+    }
+
     private synchronized void updateFinalTaskInfo(TaskInfo finalTaskInfo)
     {
-        tasksWithFinalInfo.add(finalTaskInfo.getTaskStatus().getTaskId());
+        tasksWithFinalInfo.add(finalTaskInfo.getTaskId());
         checkAllTaskFinal();
     }
 
     private synchronized void checkAllTaskFinal()
     {
         if (stateMachine.getState().isDone() && tasksWithFinalInfo.containsAll(allTasks)) {
+            if (getFragment().getStageExecutionDescriptor().isStageGroupedExecution()) {
+                // in case stage is CANCELLED/ABORTED/FAILED, number of finished lifespans can be less than total lifespans
+                checkState(finishedLifespans.size() <= totalLifespans, format("Number of finished lifespans (%s) exceeds number of total lifespans (%s)", finishedLifespans.size(), totalLifespans));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: &#39;size&#39; expression, totalLifespans. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 644. Consider formatting this data with &#39;%d&#39; instead.